docs-internal/multipart.md

```diff
@@ -0,0 +1,328 @@
+    Json.createObjectBuilder().add("alice", "bob").build());
+response.send(mp);
+```
+
+### Process a request content with the reactive model
+
+The following example consumes each part content as a `JsonObject`:
+
+```java
+req.content().as(MultiPart.class).thenAccept((multiPart) -> {
+    multiPart.subscribe(new Flow.Subscriber<BodyPart>(){
+        @Override
+        public void onSubscribe(Flow.Subscription subscription) {
+            subscription.request(Long.MAX_VALUE);
+        }
+
+        @Override
+        public void onNext(BodyPart bodyPart) {
+            bodyPart.content().as(JsonObject.class).thenAccept((str) -> {
+                System.out.println("File uploaded: " + bodyPart.headers().filename());
+            });
+        }
+
+        @Override
+        public void onError(Throwable throwable) {
+            res.status(500);
+            res.send();
+        }
+
+        @Override
+        public void onComplete() {
+            System.out.println("sending response");
+            res.send("Files uploaded successfully");
+        }
+    });
+});
+```
+
+The following example consumes each part content using reactive stream subscribers, `ServerFileWriter` is inspired from the streaming example at `webserver/examples/streaming`:
+
+```java
+req.content().as(MultiPart.class).thenAccept((multiPart) -> {
+    multiPart.subscribe(new Flow.Subscriber<BodyPart>(){
+        @Override
+        public void onSubscribe(Flow.Subscription subscription) {
+            subscription.request(Long.MAX_VALUE);
+        }
+
+        @Override
+        public void onNext(BodyPart bodyPart) {
+            bodyPart.content().subscribe(new ServerFileWriter());
+        }
+
+        @Override
+        public void onError(Throwable throwable) {
+            res.status(500);
+            res.send();
+        }
+
+        @Override
+        public void onComplete() {
+            System.out.println("sending response");
+            res.send("Files uploaded successfully");
+        }
+    });
+});
+```
+
+### Return a response with the reactive model
+
+The following example returns a multipart response using publishers of `Datachunk`, `ServerFileReader` is inspired from the streaming example at `webserver/examples/streaming`:
+
+```java
+MultiPart mp = MultiPart.builder()
+    .bodyPart(BodyPart.create(
+        BodyPartHeaders.builder().
+            .contentType(MediaType.APPLICATION_JSON)
+            .filename("file1")
+            .build(),
+        new ServerFileReader(filePath1))
+    .bodyPart(BodyPart.create(
+        BodyPartHeaders.builder().
+            .contentType(MediaType.APPLICATION_JSON)
+            .filename("file2")
+            .build(),
+        new ServerFileReader(filePath2))
+    .build();
+response.send(mp);
+```
+
+## API
+
+```java
+public interface BodyPart {
+
+    ServerRequest request();
+
+    ServerResponse response();
+
+    Content content();
+
+    BodyPartHeaders headers();
+
+    BodyPart parent();
+
```

Comment:
The use-case is for multipart/mixed where a part can itself be a multipart.

This is inspired from the bodypart entity type in jersey, http://static.javadoc.io/org.glassfish.jersey.media/jersey-media-multipart/2.27/org/glassfish/jersey/media/multipart/BodyPart.html#getParent--

Reply:
I think it's mostly to support older browser implementations of multipart/form-data. See https://tools.ietf.org/html/rfc7578#section-4.3

A MultiPart can be nested under a MultiPart, so that means that BodyParts can be nested by more than one level. Some kind tree of parts.

Having said that the `parent()` method only helps navigating the tree of parts from a single part object ; I too can't think of a valid use-case that would use that method.

Reply:
Perhaps that's enough to apply the rule of "if you don't need it don't add it" ?

Reply:
I'd drop it, it is also weird during construction of a new multi-part since it isn't clear to me when it will be set.

Reply:
I've dropped it.